python/spear/env.py

```diff
@@ -553,3 +553,96 @@ def _serialize_array(array, space, byte_order):
     assert (array <= space.high).all()
     data = array.data if byte_order is None else array.newbyteorder(byte_order).data
     return data
+
+
+# Custom Env implementation for OpenBot
+class OpenBotEnv(Env):
+
+    def __init__(self, config):
+    
+        super().__init__(config)
+        self._wheel_rotation_speeds = None
+    
+    def step(self, action):
+
+        obs, reward, is_done, step_info = super().step(action=action)
+        assert "encoder" in obs.keys()
+        self._wheel_rotation_speeds = obs["encoder"]
+
+        return obs, reward, is_done, step_info
+
+    def reset(self, reset_info=None):
+    
+        obs = super().reset(reset_info=reset_info)
+        assert "encoder" in obs.keys()
+        self._wheel_rotation_speeds = obs["encoder"]
+
+        return obs
+
+    def _apply_action(self, action):
+
+        # calculate wheel torques based on input PWM signals
+        if "apply_voltage" in action.keys():
```

Comment:
I think you should assert that `set_duty_cycle` is in `action.keys()`, the rest of the function is good. Note that you need to overwrite `self.action_space` because it's part of the public interface.